Study equality

Kris2ffer opened the pull request:
Added equals method for all internal models.
Input models are meant to be converted before testing equality.

Kris2ffer committed:
Added equals method

Kris2ffer committed:
Merge branch 'backend-updated-models' into study-equality

Kris2ffer committed:
Added equals method

Kris2ffer committed:
Add constructor and equals method

Kris2ffer committed:
Tests for equal method

Kris2ffer committed:
Added assertions for small constructors

Kris2ffer committed:
Merge branch 'master' into study-equality

Kris2ffer committed:
Removed equals check for datasets

Removed equals check for locked and loaded datasets. There is currently no conversion for datasets between input model and internal model.

gjersvik commented on RestApi/Sepes.RestApi/Model/Pod.cs:
```diff
@@ -36,10 +45,32 @@ public class Pod
+            hash.Add(allowAll);
+            hash.Add(incoming);
+            hash.Add(outgoing);
+            hash.Add(users);
```
You can't simply take the hash of lists. As that link to standard list equality. Best to not use them as part of the hash.

gjersvik commented on RestApi/Sepes.RestApi/Model/Pod.cs:
```diff
@@ -36,10 +45,32 @@ public class Pod
+            hash.Add(outgoing);
+            hash.Add(users);
+            hash.Add(locked);
+            hash.Add(loaded);
```
Do not put thing in the hash that is not in Equals. The other way around is fine. The hash rule is: If hash != hash then object must be not equal object. If hash == hash then object may be equal object.

gjersvik commented on RestApi/Sepes.RestApi/Model/Study.cs:
```diff
@@ -44,5 +47,25 @@ public class Study
+
+        public override int GetHashCode()
+        {
+            return HashCode.Combine(studyName, studyId, pods, sponsors, suppliers, archived, userIds, datasetIds);
```
Cant use pods, sponsors, suppliers, userIds or datasetIds as part of hash. See previous issue.

gjersvik commented on RestApi/Sepes.RestApi.Tests/Model/StudyTest.cs:
```diff
@@ -36,5 +37,51 @@ public void ConstructorInput()
+            Assert.True(studySmall.Equals(studySmall2));
+            Assert.Equal(1, studySmall3.pods.Count);
+            Assert.False(studySmall.Equals(studySmall3));
+        }
```
Please split up into multiple tests. This one is to big by an order of magnitude for an Equals test. There should be 3 lines in each. Create a, Create b and compare.

gjersvik reviewed (changes requested):
Some issues with hashes.

Kris2ffer committed:
Updated GetHashCode()

Kris2ffer committed:
Only testing email

Email is the important variable for now and only email will be sent from frontend.

Kris2ffer committed:
Added equals test for pod

Kris2ffer committed:
Update StudyTest.cs

Kris2ffer committed:
Added equals test

Kris2ffer committed:
Updated hashcode

gjersvik reviewed (approved):
Looks a lot better.
